File: 00-coding-101/python/basics/variables.py
# Python Variables and Data Types
# This file demonstrates basic variable usage and data types in Python

print("=== Python Variables and Data Types ===\n")

# 1. Numbers
print("1. NUMBERS:")
age = 25
height = 5.9
print(f"Age: {age} (type: {type(age)})")
print(f"Height: {height} (type: {type(height)})")
print()

# 2. Strings
print("2. STRINGS:")
name = "Alice"
message = 'Hello, World!'
multi_line = """
This is a
multi-line string
"""
print(f"Name: {name} (type: {type(name)})")
print(f"Message: {message}")
print(f"Multi-line: {multi_line}")
print()

# 3. Booleans
print("3. BOOLEANS:")
is_student = True
is_working = False
print(f"Is student: {is_student} (type: {type(is_student)})")
print(f"Is working: {is_working}")
print()

# 4. Lists (mutable)
print("4. LISTS:")
fruits = ["apple", "banana", "orange"]
numbers = [1, 2, 3, 4, 5]
mixed = [1, "hello", True, 3.14]
print(f"Fruits: {fruits} (type: {type(fruits)})")
print(f"Numbers: {numbers}")
print(f"Mixed: {mixed}")

# Adding to list
fruits.append("grape")
print(f"After adding grape: {fruits}")
print()

# 5. Tuples (immutable)
print("5. TUPLES:")
coordinates = (10, 20)
person = ("Bob", 30, "Engineer")
print(f"Coordinates: {coordinates} (type: {type(coordinates)})")
print(f"Person: {person}")
print()

# 6. Dictionaries
print("6. DICTIONARIES:")
person_info = {
    "name": "Charlie",
    "age": 35,
    "city": "New York",
    "skills": ["Python", "JavaScript", "DevOps"]
}
print(f"Person info: {person_info} (type: {type(person_info)})")
print(f"Name: {person_info['name']}")
print(f"Skills: {person_info['skills']}")
print()

# 7. Sets (unique values)
print("7. SETS:")
unique_numbers = {1, 2, 3, 3, 4, 4, 5}  # Duplicates are removed
print(f"Unique numbers: {unique_numbers} (type: {type(unique_numbers)})")
print()

# 8. Type conversion
print("8. TYPE CONVERSION:")
number_string = "42"
converted_number = int(number_string)
print(f"String '{number_string}' converted to number: {converted_number}")

float_number = 3.14
converted_int = int(float_number)
print(f"Float {float_number} converted to int: {converted_int}")

number = 123
converted_string = str(number)
print(f"Number {number} converted to string: '{converted_string}'")
print()

# 9. Variable naming conventions
print("9. VARIABLE NAMING:")
# Good naming examples
user_name = "John"
user_age = 25
is_active = True
total_count = 100

# Python naming conventions:
# - Use snake_case for variables and functions
# - Use UPPER_CASE for constants
# - Use CamelCase for classes
# - Use descriptive names

print(f"User: {user_name}, Age: {user_age}")
print(f"Active: {is_active}, Count: {total_count}")
print()

# 10. Multiple assignment
print("10. MULTIPLE ASSIGNMENT:")
x, y, z = 1, 2, 3
print(f"x={x}, y={y}, z={z}")

# Unpacking lists
colors = ["red", "green", "blue"]
color1, color2, color3 = colors
print(f"Colors: {color1}, {color2}, {color3}")

# Swapping variables
a, b = 10, 20
print(f"Before swap: a={a}, b={b}")
a, b = b, a
print(f"After swap: a={a}, b={b}")
print()

print("=== End of Variables and Data Types ===") 

File: 00-coding-101/javascript/basics/variables.js
// JavaScript Variables and Data Types
// This file demonstrates basic variable usage and data types in JavaScript

console.log("=== JavaScript Variables and Data Types ===\n");

// 1. Variable Declaration
console.log("1. VARIABLE DECLARATION:");
let age = 25;
const name = "Alice";
var oldWay = "not recommended"; // Avoid using var

console.log(`Age: ${age} (type: ${typeof age})`);
console.log(`Name: ${name} (type: ${typeof name})`);
console.log(`Old way: ${oldWay} (type: ${typeof oldWay})`);
console.log();

// 2. Numbers
console.log("2. NUMBERS:");
let integer = 42;
let float = 3.14;
let negative = -10;
let scientific = 1.23e4; // 12300

console.log(`Integer: ${integer} (type: ${typeof integer})`);
console.log(`Float: ${float} (type: ${typeof float})`);
console.log(`Negative: ${negative}`);
console.log(`Scientific: ${scientific}`);
console.log();

// 3. Strings
console.log("3. STRINGS:");
let singleQuotes = 'Hello, World!';
let doubleQuotes = "Hello, World!";
let templateLiteral = `Hello, ${name}! You are ${age} years old.`;
let multiLine = `
This is a
multi-line string
`;

console.log(`Single quotes: ${singleQuotes} (type: ${typeof singleQuotes})`);
console.log(`Double quotes: ${doubleQuotes}`);
console.log(`Template literal: ${templateLiteral}`);
console.log(`Multi-line: ${multiLine}`);
console.log();

// 4. Booleans
console.log("4. BOOLEANS:");
let isStudent = true;
let isWorking = false;

console.log(`Is student: ${isStudent} (type: ${typeof isStudent})`);
console.log(`Is working: ${isWorking}`);
console.log();

// 5. Arrays
console.log("5. ARRAYS:");
let fruits = ["apple", "banana", "orange"];
let numbers = [1, 2, 3, 4, 5];
let mixed = [1, "hello", true, 3.14];

console.log(`Fruits: ${fruits} (type: ${typeof fruits})`);
console.log(`Numbers: ${numbers}`);
console.log(`Mixed: ${mixed}`);

// Adding to array
fruits.push("grape");
console.log(`After adding grape: ${fruits}`);

// Array methods
console.log(`First fruit: ${fruits[0]}`);
console.log(`Array length: ${fruits.length}`);
console.log();

// 6. Objects
console.log("6. OBJECTS:");
let person = {
    name: "Bob",
    age: 30,
    city: "New York",
    skills: ["JavaScript", "Python", "DevOps"]
};

console.log(`Person: ${JSON.stringify(person)} (type: ${typeof person})`);
console.log(`Name: ${person.name}`);
console.log(`Age: ${person.age}`);
console.log(`Skills: ${person.skills}`);

// Adding properties
person.job = "Developer";
console.log(`After adding job: ${JSON.stringify(person)}`);
console.log();

// 7. Undefined and Null
console.log("7. UNDEFINED AND NULL:");
let undefinedVar;
let nullVar = null;

console.log(`Undefined: ${undefinedVar} (type: ${typeof undefinedVar})`);
console.log(`Null: ${nullVar} (type: ${typeof nullVar})`);
console.log();

// 8. Type conversion
console.log("8. TYPE CONVERSION:");
let numberString = "42";
let convertedNumber = parseInt(numberString);
let convertedFloat = parseFloat("3.14");

console.log(`String '${numberString}' converted to number: ${convertedNumber}`);
console.log(`String '3.14' converted to float: ${convertedFloat}`);

let number = 123;
let convertedString = number.toString();
console.log(`Number ${number} converted to string: '${convertedString}'`);

// Automatic type conversion
console.log(`"5" + 3 = ${"5" + 3}`); // String concatenation
console.log(`"5" - 3 = ${"5" - 3}`); // Number subtraction
console.log();

// 9. Template literals and string interpolation
console.log("9. TEMPLATE LITERALS:");
let firstName = "John";
let lastName = "Doe";
let fullName = `${firstName} ${lastName}`;
let message = `Hello, ${fullName}! Welcome to JavaScript.`;

console.log(`Full name: ${fullName}`);
console.log(`Message: ${message}`);
console.log();

// 10. Destructuring
console.log("10. DESTRUCTURING:");
let colors = ["red", "green", "blue"];
let [firstColor, secondColor, thirdColor] = colors;
console.log(`Colors: ${firstColor}, ${secondColor}, ${thirdColor}`);

let user = {
    username: "johndoe",
    email: "john@example.com",
    role: "admin"
};
let { username, email, role } = user;
console.log(`User: ${username}, Email: ${email}, Role: ${role}`);
console.log();

// 11. Spread operator
console.log("11. SPREAD OPERATOR:");
let arr1 = [1, 2, 3];
let arr2 = [...arr1, 4, 5];
console.log(`Original array: ${arr1}`);
console.log(`Spread array: ${arr2}`);

let obj1 = { x: 1, y: 2 };
let obj2 = { ...obj1, z: 3 };
console.log(`Original object: ${JSON.stringify(obj1)}`);
console.log(`Spread object: ${JSON.stringify(obj2)}`);
console.log();

// 12. Variable naming conventions
console.log("12. VARIABLE NAMING:");
// Good naming examples
let userName = "John";
let userAge = 25;
let isActive = true;
let totalCount = 100;

// JavaScript naming conventions:
// - Use camelCase for variables and functions
// - Use UPPER_CASE for constants
// - Use PascalCase for classes
// - Use descriptive names

console.log(`User: ${userName}, Age: ${userAge}`);
console.log(`Active: ${isActive}, Count: ${totalCount}`);
console.log();

console.log("=== End of Variables and Data Types ==="); 